File: php/ant_ffi/src/Types/Network.php
<?php

declare(strict_types=1);

namespace AntFfi\Types;

use AntFfi\AntFfiException;
use AntFfi\FFILoader;
use AntFfi\NativeHandle;
use AntFfi\RustBuffer;
use FFI;
use FFI\CData;

/**
 * Network configuration.
 */
final class Network extends NativeHandle
{
    /**
     * Create a local network configuration.
     */
    public static function local(): self
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $handle = $ffi->uniffi_ant_ffi_fn_constructor_network_new(
            1, // true for local
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new self($handle);
    }

    /**
     * Create a mainnet network configuration.
     */
    public static function mainnet(): self
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $handle = $ffi->uniffi_ant_ffi_fn_constructor_network_new(
            0, // false for mainnet
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new self($handle);
    }

    /**
     * Create a custom network configuration.
     */
    public static function custom(string $rpcUrl, string $paymentTokenAddress, string $dataPaymentsAddress): self
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');

        $rpcBuffer = RustBuffer::fromString($rpcUrl);
        $tokenBuffer = RustBuffer::fromString($paymentTokenAddress);
        $paymentsBuffer = RustBuffer::fromString($dataPaymentsAddress);

        $handle = $ffi->uniffi_ant_ffi_fn_constructor_network_custom(
            $rpcBuffer,
            $tokenBuffer,
            $paymentsBuffer,
            FFI::addr($status)
        );

        RustBuffer::checkStatus($status);

        return new self($handle);
    }

    protected function freeHandle(): void
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        $ffi->uniffi_ant_ffi_fn_free_network($this->handle, FFI::addr($status));
    }

    protected function cloneHandle(): CData
    {
        $ffi = FFILoader::get();
        $status = $ffi->new('RustCallStatus');
        return $ffi->uniffi_ant_ffi_fn_clone_network($this->handle, FFI::addr($status));
    }
}

File: dart/ant_ffi/lib/src/types/network.dart
import 'dart:ffi';
import 'package:ffi/ffi.dart';
import '../native/bindings.dart';
import '../native/library.dart';
import '../native/rust_buffer.dart';

late final _bindings = AntFfiBindings(antFfiLib);

/// Network configuration for connecting to Autonomi.
///
/// Use [Network.local] for local testnet or [Network.mainnet] for production.
class Network {
  final Pointer<Void> _handle;
  bool _disposed = false;

  Network._(this._handle);

  /// Clones the internal handle for passing to FFI.
  /// Required because UniFFI consumes one Arc reference per call.
  Pointer<Void> _clone() {
    _checkNotDisposed();
    final status = calloc<RustCallStatus>();
    try {
      final cloned = _bindings.uniffi_ant_ffi_fn_clone_network(_handle, status);
      _checkStatus(status.ref);
      return cloned;
    } finally {
      calloc.free(status);
    }
  }

  /// Creates a network configuration for local testnet.
  factory Network.local() {
    final status = calloc<RustCallStatus>();
    try {
      final handle = _bindings.uniffi_ant_ffi_fn_constructor_network_new(1, status);
      _checkStatus(status.ref);
      return Network._(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Creates a network configuration for mainnet.
  factory Network.mainnet() {
    final status = calloc<RustCallStatus>();
    try {
      final handle = _bindings.uniffi_ant_ffi_fn_constructor_network_new(0, status);
      _checkStatus(status.ref);
      return Network._(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Creates a custom network configuration.
  factory Network.custom({
    required String rpcUrl,
    required String paymentTokenAddress,
    required String dataPaymentsAddress,
    String? royaltiesPkHex,
  }) {
    final status = calloc<RustCallStatus>();
    try {
      final rpcBuffer = stringToRustBuffer(rpcUrl);
      final tokenBuffer = stringToRustBuffer(paymentTokenAddress);
      final dataBuffer = stringToRustBuffer(dataPaymentsAddress);
      final royaltiesBuffer = optionStringToRustBuffer(royaltiesPkHex);
      final handle = _bindings.uniffi_ant_ffi_fn_constructor_network_custom(
        rpcBuffer,
        tokenBuffer,
        dataBuffer,
        royaltiesBuffer,
        status,
      );
      _checkStatus(status.ref);
      return Network._(handle);
    } finally {
      calloc.free(status);
    }
  }

  /// Returns the internal handle for FFI calls.
  Pointer<Void> get handle => _clone();

  /// Releases the native resources associated with this network.
  void dispose() {
    if (_disposed) return;
    _disposed = true;
    final status = calloc<RustCallStatus>();
    try {
      _bindings.uniffi_ant_ffi_fn_free_network(_handle, status);
    } finally {
      calloc.free(status);
    }
  }

  void _checkNotDisposed() {
    if (_disposed) {
      throw StateError('Network has been disposed');
    }
  }
}

void _checkStatus(RustCallStatus status) {
  if (status.code != 0) {
    String errorMessage = 'Network operation failed with code ${status.code}';
    if (status.errorBuf.len > 0) {
      try {
        errorMessage = rustBufferToStringWithPrefix(status.errorBuf);
      } catch (_) {
        try {
          errorMessage = rustBufferToString(status.errorBuf);
        } catch (_) {}
      }
    }
    throw AntFfiException(errorMessage, status.code);
  }
}
